cleanlab/object_detection/rank.py

```diff
@@ -0,0 +1,1040 @@
+        the (x2, y2) position is at the bottom right corner
+    bb2 : dict
+        Keys: {'x1', 'x2', 'y1', 'y2'}
+        The (x, y) position is at the top left corner,
+        the (x2, y2) position is at the bottom right corner
+    Returns
+    -------
+    float
+        in [0, 1]
+    """
+    assert bb1["x1"] < bb1["x2"]
+    assert bb1["y1"] < bb1["y2"]
+    assert bb2["x1"] < bb2["x2"]
+    assert bb2["y1"] < bb2["y2"]
+
+    # determine the coordinates of the intersection rectangle
+    x_left = max(bb1["x1"], bb2["x1"])
+    y_top = max(bb1["y1"], bb2["y1"])
+    x_right = min(bb1["x2"], bb2["x2"])
+    y_bottom = min(bb1["y2"], bb2["y2"])
+
+    if x_right < x_left or y_bottom < y_top:
+        return 0.0
+
+    # The intersection of two axis-aligned bounding boxes is always an
+    # axis-aligned bounding box
+    intersection_area = (x_right - x_left) * (y_bottom - y_top)
+
+    # compute the area of both AABBs
+    bb1_area = (bb1["x2"] - bb1["x1"]) * (bb1["y2"] - bb1["y1"])
+    bb2_area = (bb2["x2"] - bb2["x1"]) * (bb2["y2"] - bb2["y1"])
+    assert intersection_area - 0.1 <= bb1_area
+    assert intersection_area - 0.1 <= bb2_area
+
+    # compute the intersection over union by taking the intersection
+    # area and dividing it by the sum of prediction + ground-truth
+    # areas - the interesection area
+    iou = intersection_area / float(bb1_area + bb2_area - intersection_area)
+    # There are some hyper-parameters here like consider tile area/object area
+    assert iou >= 0.0
+    assert iou - 0.01 <= 1.0
+    return iou
+
+
+def _euc_dis(box1: List[float], box2: List[float]) -> float:
+    """Calculates the euclidian distance between `box1` and `box2`."""
+    x1, y1 = (box1[0] + box1[2]) / 2, (box1[1] + box1[3]) / 2
+    x2, y2 = (box2[0] + box2[2]) / 2, (box2[1] + box2[3]) / 2
+    p1 = np.array([x1, y1])
+    p2 = np.array([x2, y2])
+    val2 = np.exp(-np.linalg.norm(p1 - p2) * EUC_FACTOR)
+    return val2
+
+
+def _get_dist_matrix(bb1_list: np.ndarray, bb2_list: np.ndarray) -> np.ndarray:
+    """Returns a distance matrix of distances from all of boxes in bb1_list to all of boxes in bb2_list."""
+    wd = np.zeros(shape=(len(bb1_list), len(bb2_list)))
+    for i in range(len(bb1_list)):
+        for j in range(len(bb2_list)):
+            wd[i][j] = _euc_dis(bb1_list[i], bb2_list[j])
+    return wd
+
+
+def _softmax(x: np.ndarray, temperature: float = 0.99, axis: int = 0) -> np.ndarray:
+    """Gets softmax of scores."""
+    x = x / temperature
+    x_max = np.amax(x, axis=axis, keepdims=True)
+    exp_x_shifted = np.exp(x - x_max)
+    return exp_x_shifted / np.sum(exp_x_shifted, axis=axis, keepdims=True)
+
+
+def _softmin1D(scores: np.ndarray, temperature: float = 0.99, axis: int = 0) -> float:
+    """Returns softmin of passed in scores."""
+    scores = np.array(scores)
+    softmax_scores = _softmax(-1 * scores, temperature, axis)
+    return np.dot(softmax_scores, scores)
```

Comment:
consider moving this file to cleanlab.internal in some utils file, where it can be more easily be re-used across the cleanlab library. And then remove the underscore name

Reply:
also move the _softmax() function to the same utils file and remove the underscore. 

In a follow-up PR (please don't forget): update the softmax used in token-and multilabel classification to both use this new more-stable softmax method